tests/PHPStanTests/nsrt/preg-match.php

```diff
@@ -15,25 +15,33 @@ function doMatch(string $s): void
     assertType('array{}|array{string}', $matches);
 
     if (Preg::match('/Price: (£|€)\d+/', $s, $matches)) {
-        assertType('array{string, string|null}', $matches);
+        assertType('array{string, string}', $matches);
     } else {
         assertType('array{}', $matches);
     }
-    assertType('array{}|array{string, string|null}', $matches);
+    assertType('array{}|array{string, string}', $matches);
 
     if (Preg::match('/Price: (£|€)?\d+/', $s, $matches)) {
-        assertType('array{0: string, 1?: string|null}', $matches);
+        assertType('array{0: string, 1: string|null}', $matches);
     } else {
         assertType('array{}', $matches);
     }
-    assertType('array{}|array{0: string, 1?: string|null}', $matches);
+    assertType('array{}|array{0: string, 1: string|null}', $matches);
+
+    // passing the PREG_UNMATCHED_AS_NULL should change nothing compared to above as it is always set
+    if (Preg::match('/Price: (£|€)?\d+/', $s, $matches, PREG_UNMATCHED_AS_NULL)) {
+        assertType('array{0: string, 1: string|null}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+    assertType('array{}|array{0: string, 1: string|null}', $matches);
```

Comment:
Both of these are failing right now.. and I'm not sure why.

Reply:
the `(£|€)?` is a optional group, therefore the assertions looks wrong to me. 

IMO it should be 
`assertType('array{}|array{0: string, 1?: string|null}', $matches);`

(and thats whats the extension calculates)

Reply:
But with PREG_UNMATCHED_AS_NULL the offset 1 will always exist. It is correct that it's nullable because it's optional though. https://3v4l.org/VeDob

Reply:
I see. bug on my end ;) - working on a fix

Reply:
we both were right. it depends on the PHP version ;-)

https://github.com/phpstan/phpstan-src/pull/3229

Reply:
yes php7.2/3 have a borked null handling except in preg match all